pkg/cache/cache.go

```diff
@@ -117,38 +118,46 @@ func (c *Cache) WaitForPodsReady(ctx context.Context) {
 		return
 	}
 
-	c.Lock()
-	defer c.Unlock()
-
 	log := ctrl.LoggerFrom(ctx)
+
+	podsReadyCond := func(ch chan struct{}) {
```

Comment:
yes, the main problem is in the default case in a select statement, it becomes non-blocking and when context is canceled exactly after of passing the first condition of select, but before calling the function 'Wait'. If we put the channel instead of default case we'll wait until it could receive 'close' from podsReadyCondCh.